sanic/asgi.py

```diff
@@ -0,0 +1,345 @@
+        self.asgi_app = asgi_app
+
+        if "before_server_start" in self.asgi_app.sanic_app.listeners:
+            warnings.warn(
+                'You have set a listener for "before_server_start" '
+                "in ASGI mode. "
+                "It will be executed as early as possible, but not before "
+                "the ASGI server is started."
+            )
+        if "after_server_stop" in self.asgi_app.sanic_app.listeners:
+            warnings.warn(
+                'You have set a listener for "after_server_stop" '
+                "in ASGI mode. "
+                "It will be executed as late as possible, but not after "
+                "the ASGI server is stopped."
+            )
+
+    async def pre_startup(self) -> None:
+        for handler in self.asgi_app.sanic_app.listeners[
+            "before_server_start"
+        ]:
+            response = handler(
+                self.asgi_app.sanic_app, self.asgi_app.sanic_app.loop
+            )
+            if isawaitable(response):
+                await response
+
+    async def startup(self) -> None:
+        for handler in self.asgi_app.sanic_app.listeners[
+            "before_server_start"
+        ]:
+            response = handler(
+                self.asgi_app.sanic_app, self.asgi_app.sanic_app.loop
+            )
+            if isawaitable(response):
+                await response
+
+        for handler in self.asgi_app.sanic_app.listeners["after_server_start"]:
+            response = handler(
+                self.asgi_app.sanic_app, self.asgi_app.sanic_app.loop
+            )
+            if isawaitable(response):
+                await response
+
+    async def shutdown(self) -> None:
+        for handler in self.asgi_app.sanic_app.listeners["before_server_stop"]:
+            response = handler(
+                self.asgi_app.sanic_app, self.asgi_app.sanic_app.loop
+            )
+            if isawaitable(response):
+                await response
+
+        for handler in self.asgi_app.sanic_app.listeners["after_server_stop"]:
+            response = handler(
+                self.asgi_app.sanic_app, self.asgi_app.sanic_app.loop
+            )
+            if isawaitable(response):
+                await response
+
+    async def __call__(
+        self, scope: ASGIScope, receive: ASGIReceive, send: ASGISend
+    ) -> None:
+        message = await receive()
+        if message["type"] == "lifespan.startup":
+            await self.startup()
+            await send({"type": "lifespan.startup.complete"})
+
+        message = await receive()
+        if message["type"] == "lifespan.shutdown":
+            await self.shutdown()
+            await send({"type": "lifespan.shutdown.complete"})
+
+
+class ASGIApp:
+    def __init__(self) -> None:
+        self.ws = None
+
+    @classmethod
+    async def create(
+        cls, sanic_app, scope: ASGIScope, receive: ASGIReceive, send: ASGISend
+    ) -> "ASGIApp":
```

Comment:
double quote ?

Reply:
On the annotation? Because we are also supporting 3.6, we cannot use `from __future__ import annotations`. And, in order to avoid a circular reference, I opted for a string type annotation here.